module/executiondatasync/optimistic_syncing/execution_result_query_provider_test.go

```diff
@@ -0,0 +1,298 @@
+	})
+
+	suite.Run("required executors not found returns error", func() {
+		receipts := make(flow.ExecutionReceiptList, totalReceipts)
+		for i := 0; i < totalReceipts; i++ {
+			r := unittest.ReceiptForBlockFixture(&block)
+			r.ExecutorID = allExecutionNodes[i].NodeID
+			r.ExecutionResult = *executionResult
+			receipts[i] = r
+		}
+
+		suite.receipts.On("ByBlockID", block.ID()).Return(receipts, nil)
+		suite.setupIdentitiesMock(allExecutionNodes)
+
+		// Require executors that didn't produce any receipts
+		nonExistentExecutors := unittest.IdentityListFixture(2, unittest.WithRole(flow.RoleExecution)).NodeIDs()
+		criteria := Criteria{
+			AgreeingExecutors: 2,
+			RequiredExecutors: nonExistentExecutors,
+		}
+
+		_, err := suite.provider.ExecutionResultQuery(block.ID(), criteria)
+		suite.Require().Error(err)
+
+		suite.Assert().True(backend.IsInsufficientExecutionReceipts(err))
+	})
+}
+
+// TestRootBlockHandling tests the special case handling for root blocks.
+func (suite *ExecutionResultQueryProviderSuite) TestRootBlockHandling() {
+	allExecutionNodes := unittest.IdentityListFixture(5, unittest.WithRole(flow.RoleExecution))
+	suite.setupIdentitiesMock(allExecutionNodes)
+
+	suite.Run("root block returns execution nodes without execution result", func() {
+		query, err := suite.provider.ExecutionResultQuery(suite.rootBlock.ID(), Criteria{})
+		suite.Require().NoError(err)
+
+		suite.Assert().Equal(suite.rootBlockResult, query.ExecutionResult)
+		suite.Assert().Len(query.ExecutionNodes.NodeIDs(), maxNodesCnt)
+		suite.Assert().Subset(allExecutionNodes.NodeIDs(), query.ExecutionNodes.NodeIDs())
+	})
+
+	suite.Run("root block with required executors", func() {
+		requiredExecutors := allExecutionNodes[0:2].NodeIDs()
+		criteria := Criteria{
+			AgreeingExecutors: 1,
+			RequiredExecutors: requiredExecutors,
+		}
+
+		query, err := suite.provider.ExecutionResultQuery(suite.rootBlock.ID(), criteria)
+		suite.Require().NoError(err)
+
+		suite.Assert().Equal(suite.rootBlockResult, query.ExecutionResult)
+		suite.Assert().ElementsMatch(query.ExecutionNodes.NodeIDs(), requiredExecutors)
+	})
+}
+
+// TestPreferredAndRequiredExecutionNodes tests the interaction with preferred and required execution nodes.
+func (suite *ExecutionResultQueryProviderSuite) TestPreferredAndRequiredExecutionNodes() {
+	block := unittest.BlockFixture()
+	allExecutionNodes := unittest.IdentityListFixture(6, unittest.WithRole(flow.RoleExecution))
+	executionResult := unittest.ExecutionResultFixture()
+
+	// Create receipts from the first 4 execution nodes
+	receipts := make(flow.ExecutionReceiptList, 4)
+	for i := 0; i < 4; i++ {
+		r := unittest.ReceiptForBlockFixture(&block)
+		r.ExecutorID = allExecutionNodes[i].NodeID
+		r.ExecutionResult = *executionResult
+		receipts[i] = r
+	}
+
+	suite.receipts.On("ByBlockID", block.ID()).Return(receipts, nil)
+	suite.setupIdentitiesMock(allExecutionNodes)
+
+	suite.Run("with preferred execution nodes", func() {
```

Comment:
I think we need to test 5 configurations
* user required is set
* only preferred execution nodes is set
* only required execution nodes is set
* both are set
* neither are set

some have different conditions (e.g. preferred could either have no matches, partial match, or complete match